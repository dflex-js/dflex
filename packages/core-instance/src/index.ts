export type {
  INodeAbstract as IAbstract,
  CoreInput,
  TransitionHistory,
  INode,
} from "./node";

export { Node } from "./node";
