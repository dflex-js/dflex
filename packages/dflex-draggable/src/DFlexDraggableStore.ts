import type { Keys } from "@dflex/dom-gen";
import DFlexBaseStore from "@dflex/store";
import { canUseDOM } from "@dflex/utils";

declare global {
  // eslint-disable-next-line
  var $DFlex_Draggable: DFlexDraggableStore;
}

class DFlexDraggableStore extends DFlexBaseStore {
  constructor() {
    super();
    this._initBranch = this._initBranch.bind(this);
    this._initElmDOMInstance = this._initElmDOMInstance.bind(this);
  }

  private _initElmDOMInstance(id: string) {
    const [dflexNode, DOM] = this.getElmWithDOM(id);

    dflexNode.resume(DOM);
  }

  private _initBranch(keys: Keys) {
    const { SK } = keys;

    this.getElmBranchByKey(SK).forEach(this._initElmDOMInstance);
  }

  /**
   * Register element for Draggable store.
   * @param id
   */
  // @ts-ignore
  register(id: string) {
    super.register(
      {
        id,
        depth: 0,
        readonly: false,
      },
      this._initBranch
    );
  }
}

export default (function createStoreInstance() {
  const store = new DFlexDraggableStore();

  if (__DEV__) {
    if (canUseDOM()) {
      if (!globalThis.$DFlex_Draggable) {
        globalThis.$DFlex_Draggable = store;
      }
    }
  }

  return store;
})();
