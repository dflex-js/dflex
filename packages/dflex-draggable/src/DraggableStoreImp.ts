import DFlexBaseStore from "@dflex/store";

class DraggableStoreImp extends DFlexBaseStore {
  /**
   * Register element for Draggable store
   */
  // @ts-expect-error
  register(id: string) {
    super.register({
      id,
      depth: 0,
      parentID: "",
      readonly: false,
    });
  }
}

export default (function createStoreInstance() {
  const store = new DraggableStoreImp();

  return store;
})();
