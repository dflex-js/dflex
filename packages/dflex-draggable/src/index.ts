export { default as Draggable } from "./DFlexDraggable";
export { default as store } from "./DFlexDraggableStore";
export { default as DFlexBaseDraggable } from "./DFlexBaseDraggable";
