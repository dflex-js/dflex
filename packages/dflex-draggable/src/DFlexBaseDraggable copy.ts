/* eslint-disable no-dupe-class-members */
/* eslint-disable no-unused-vars */
import { DFlexBaseElement } from "@dflex/core-instance";
import { PointNum, getSelection, rmEmptyAttr } from "@dflex/utils";
import type { AxesPoint } from "@dflex/utils";

const MIRROR_STYLE = {
  POSITION: "position",
  TOP: "top",
  LEFT: "left",
  WIDTH: "width",
  HEIGHT: "height",
  Z_INDEX: "z-index",
  MARGIN: "margin",
};

const ORIGIN_STYLE = {
  POSITION: "position",
  Z_INDEX: "z-index",
};

const USER_SELECT_STYLE = {
  PROPERTY: "user-select",
  VALUE: "none",
};

if (__DEV__) {
  Object.freeze(MIRROR_STYLE);
  Object.freeze(ORIGIN_STYLE);
  Object.freeze(USER_SELECT_STYLE);
}

function setMirrorStyle(
  mirrorStyle: CSSStyleDeclaration,
  viewportPos: [number, number] | null,
  dimensions: PointNum | null
): void {
  let width = "auto";
  let height = "auto";

  if (dimensions) {
    if (dimensions.y > 0) {
      height = `${dimensions.y}px`;
    }

    if (dimensions.x > 0) {
      width = `${dimensions.x}px`;
    }
  }

  const [top = 0, left = 0] = viewportPos || [];

  mirrorStyle.cssText = `
    ${MIRROR_STYLE.POSITION}: fixed;
    ${MIRROR_STYLE.TOP}: ${top}px;
    ${MIRROR_STYLE.LEFT}: ${left}px;
    ${MIRROR_STYLE.WIDTH}: ${width};
    ${MIRROR_STYLE.HEIGHT}: ${height};
    ${MIRROR_STYLE.Z_INDEX}: 99;
    ${MIRROR_STYLE.MARGIN}: 0;
  `;
}

function setOrUnsetOriginStyle(
  originStyle: CSSStyleDeclaration,
  shouldSet: boolean
) {
  if (shouldSet) {
    originStyle.cssText = `
    ${ORIGIN_STYLE.POSITION}: relative;
    ${ORIGIN_STYLE.Z_INDEX}: 99;
  `;

    return;
  }

  originStyle.removeProperty(ORIGIN_STYLE.POSITION);
  originStyle.removeProperty(ORIGIN_STYLE.Z_INDEX);
}

function disableUserSelect(): void {
  document.body.style.setProperty(
    USER_SELECT_STYLE.PROPERTY,
    USER_SELECT_STYLE.VALUE
  );

  const domSelection = getSelection();

  if (domSelection) {
    domSelection.removeAllRanges();
  }
}

function restoreUserSelect() {
  document.body.style.removeProperty(USER_SELECT_STYLE.PROPERTY);
}

class DFlexBaseDraggable<T extends DFlexBaseElement> {
  draggedElm: T;

  draggedDOM: HTMLElement;

  /**
   * When dragging start, element shouldn't jump from its translate. So, we
   * calculate offset that make translate X,Y start from zero:
   *  goToX = x + this.outerOffsetX.
   *  goToY = y + this.outerOffsetY.
   *
   * goToX and goToY both should be zero with first click. Starts with simple
   * equating: initX = X. Taking into considerations translate value.
   *
   */
  private _outerOffset: PointNum;

  private _restoreContentEditable: boolean;

  private _preservedAriaDisabled: string | null;

  protected translatePlaceholder: PointNum;

  /**
   * Creates an instance of AbstractDraggable.
   * Works Only on dragged element level.
   *
   * @param abstractCoreElm -
   * @param DOM -
   * @param initCoordinates -
   */
  constructor(
    abstractCoreElm: T,
    DOM: HTMLElement,
    initCoordinates: AxesPoint
  ) {
    this.draggedElm = abstractCoreElm;
    this.draggedDOM = DOM;

    const { translate } = this.draggedElm;

    this._outerOffset = new PointNum(
      -initCoordinates.x + translate.x,
      -initCoordinates.y + translate.y
    );

    this.translatePlaceholder = new PointNum(0, 0);
    this._restoreContentEditable = false;
    this._preservedAriaDisabled = null;
  }

  /**
   *
   * @param originDOM
   * @param mirrorDOM
   * @param isAddingProps
   * @param dimensions
   * @param viewportPos
   * @returns
   */
  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement,
    isAddingProps: true,
    isMigratedInScroll: false,
    dimensions: PointNum,
    viewportPos: [number, number]
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement,
    isAddingProps: false,
    isMigratedInScroll: true,
    dimensions: PointNum,
    viewportPos: null
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement,
    isAddingProps: false,
    isMigratedInScroll: false,
    dimensions: null,
    viewportPos: null
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: null,
    isAddingProps: true,
    isMigratedInScroll: false,
    dimensions: null,
    viewportPos: null
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: null,
    isAddingProps: false,
    isMigratedInScroll: false,
    dimensions: null,
    viewportPos: null
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement | null,
    isAddingProps: boolean,
    isMigratedInScroll: boolean,
    dimensions: PointNum | null,
    viewportPos: [number, number] | null
  ): void {
    const { style: originStyle } = originDOM;

    if (isAddingProps) {
      this._preservedAriaDisabled = originDOM.ariaDisabled;
      originDOM.ariaDisabled = "true";
      this.draggedElm.setAttribute(originDOM, "DRAGGED", "true");

      disableUserSelect();

      if (mirrorDOM !== null) {
        const { style: mirrorStyle } = mirrorDOM;

        mirrorDOM.ariaLabel = "Draggable";
        mirrorDOM.id = `dflex-draggable-mirror__${originDOM.id}`;

        delete mirrorDOM.dataset.index;

        setMirrorStyle(mirrorStyle, viewportPos, dimensions);

        originStyle.setProperty("opacity", "0");

        return;
      }

      originDOM.ariaLabel = "Draggable";

      setOrUnsetOriginStyle(originStyle, true);

      return;
    }

    restoreUserSelect();

    if (this._restoreContentEditable) {
      originDOM.contentEditable = "true";
    }

    originDOM.ariaLabel = null;
    originDOM.ariaDisabled = this._preservedAriaDisabled;

    this.draggedElm.clearAttributes(originDOM);

    if (mirrorDOM !== null) {
      if (isMigratedInScroll) {
        originStyle.setProperty("position", "absolute");
        if (dimensions!.x > 0) {
          originStyle.setProperty("width", `${dimensions!.x}px`);
        }
        if (dimensions!.y > 0) {
          originStyle.setProperty("height", `${dimensions!.y}px`);
        }
      }

      originStyle.removeProperty("opacity");

      mirrorDOM.remove();
    } else {
      setOrUnsetOriginStyle(originStyle, false);
    }

    rmEmptyAttr(originDOM, "style");
    rmEmptyAttr(document.body, "style");
  }

  /**
   * Executes dragging by applying transform.
   * Writes to draggedElmCurrentOffset in Transform class.
   * Set values to isDragged flags.
   *
   * @param x - mouse x coordinates
   * @param y - mouse y coordinates
   */
  protected translate(x: number, y: number) {
    /**
     * Calculates translate coordinates.
     *
     * Indicates dragged y-transformation that's will be updated during the
     * dropping process. Updating Y immediately will effect calculations in
     * transform, that's why it is updated when dragging is done.
     */
    this.translatePlaceholder.setAxes(
      x + this._outerOffset.x,
      y + this._outerOffset.y
    );

    DFlexBaseElement.transform(
      this.draggedDOM,
      this.translatePlaceholder.x,
      this.translatePlaceholder.y
    );
  }
}

export default DFlexBaseDraggable;
