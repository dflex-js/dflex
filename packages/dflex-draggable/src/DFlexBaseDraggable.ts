/* eslint-disable no-dupe-class-members */
/* eslint-disable no-unused-vars */
import { DFlexBaseElement } from "@dflex/core-instance";
import {
  PointNum,
  getSelection,
  removeStyleProperty,
  setStyleProperty,
  updateDOMAttr,
} from "@dflex/utils";
import type { AxesPoint } from "@dflex/utils";

const MIRROR_ID_PREFIX = "dflex-draggable-mirror";

function updatedDraggedAttr(originDOM: HTMLElement, isRemove: boolean): void {
  updateDOMAttr(originDOM, "dragged", isRemove, false);
}

function setMirrorStyle(
  mirrorDOM: HTMLElement,
  viewportPos: [number, number] | null,
  dimensions: PointNum | null,
): void {
  const [top = 0, left = 0] = viewportPos || [];

  setStyleProperty(mirrorDOM, "position", "fixed");
  setStyleProperty(mirrorDOM, "top", `${top}px`);
  setStyleProperty(mirrorDOM, "left", `${left}px`);

  if (dimensions) {
    if (dimensions.y > 0) {
      setStyleProperty(mirrorDOM, "height", `${dimensions.y}px`);
    }

    if (dimensions.x > 0) {
      setStyleProperty(mirrorDOM, "width", `${dimensions.x}px`);
    }
  }

  setStyleProperty(mirrorDOM, "z-index", "99");
  setStyleProperty(mirrorDOM, "margin", "0");
}

function setOrUnsetOriginStyle(originDOM: HTMLElement, shouldSet: boolean) {
  if (shouldSet) {
    setStyleProperty(originDOM, "position", "relative");
    setStyleProperty(originDOM, "z-index", "99");
  } else {
    removeStyleProperty(originDOM, "position");
    removeStyleProperty(originDOM, "z-index");
  }
}

function rmEmptyAttr(DOM: HTMLElement, attribute: string) {
  if (!DOM.hasAttribute(attribute)) {
    return;
  }

  const value = DOM.getAttribute(attribute);

  if (value && value.trim() === "") {
    DOM.removeAttribute(attribute);
  }
}

class DFlexBaseDraggable<T extends DFlexBaseElement> {
  draggedElm: T;

  draggedDOM: HTMLElement;

  /**
   * When dragging start, element shouldn't jump from its translate. So, we
   * calculate offset that make translate X,Y start from zero:
   *  goToX = x + this.outerOffsetX.
   *  goToY = y + this.outerOffsetY.
   *
   * goToX and goToY both should be zero with first click. Starts with simple
   * equating: initX = X. Taking into considerations translate value.
   *
   */
  private _outerOffset: PointNum;

  private _preservedAriaDisabled: string | null;

  protected translatePlaceholder: PointNum;

  /**
   * Creates an instance of AbstractDraggable.
   * Works Only on dragged element level.
   *
   * @param abstractCoreElm -
   * @param DOM -
   * @param initCoordinates -
   */
  constructor(
    abstractCoreElm: T,
    DOM: HTMLElement,
    initCoordinates: AxesPoint,
  ) {
    this.draggedElm = abstractCoreElm;
    this.draggedDOM = DOM;

    const { translate } = this.draggedElm;

    this._outerOffset = new PointNum(
      translate.x - initCoordinates.x,
      translate.y - initCoordinates.y,
    );

    this.translatePlaceholder = new PointNum(0, 0);

    this._preservedAriaDisabled = null;
  }

  private _preserveDOMAttributes(DOM: HTMLElement, shouldSet: boolean): void {
    if (shouldSet) {
      this._preservedAriaDisabled = DOM.ariaDisabled;

      DOM.ariaDisabled = "true";

      return;
    }

    DOM.ariaDisabled = this._preservedAriaDisabled;
  }

  /**
   *
   * @param originDOM
   * @param mirrorDOM
   * @param isAddingProps
   * @param dimensions
   * @param viewportPos
   * @returns
   */
  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement,
    isAddingProps: true,
    dimensions: PointNum,
    viewportPos: [number, number],
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: null,
    isAddingProps: true,
    dimensions: null,
    viewportPos: null,
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement,
    isAddingProps: false,
    dimensions: null,
    viewportPos: null,
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: null,
    isAddingProps: false,
    dimensions: null,
    viewportPos: null,
  ): void;

  protected setDOMAttrAndStyle(
    originDOM: HTMLElement,
    mirrorDOM: HTMLElement | null,
    isAddingProps: boolean,
    dimensions: PointNum | null,
    viewportPos: [number, number] | null,
  ): void {
    if (isAddingProps) {
      this._preserveDOMAttributes(originDOM, true);

      updatedDraggedAttr(originDOM, false);

      if (mirrorDOM !== null) {
        mirrorDOM.ariaLabel = "Draggable";

        mirrorDOM.id = `${MIRROR_ID_PREFIX}_${originDOM.id}`;

        delete mirrorDOM.dataset.index;

        setMirrorStyle(mirrorDOM, viewportPos, dimensions);

        setStyleProperty(originDOM, "opacity", "0");
      } else {
        originDOM.ariaLabel = "Draggable";

        setOrUnsetOriginStyle(originDOM, true);
      }

      setStyleProperty(document.body, "user-select", "none");

      const domSelection = getSelection();

      if (domSelection) {
        domSelection.removeAllRanges();
      }

      return;
    }

    removeStyleProperty(document.body, "user-select");

    this._preserveDOMAttributes(originDOM, false);

    originDOM.ariaLabel = null;

    updatedDraggedAttr(originDOM, true);

    if (mirrorDOM !== null) {
      removeStyleProperty(originDOM, "opacity");

      mirrorDOM.remove();
    } else {
      setOrUnsetOriginStyle(originDOM, false);
    }

    rmEmptyAttr(originDOM, "style");
    rmEmptyAttr(document.body, "style");
  }

  /**
   * Executes dragging by applying transform.
   * Writes to draggedElmCurrentOffset in Transform class.
   * Set values to isDragged flags.
   *
   * @param x - mouse x coordinates
   * @param y - mouse y coordinates
   */
  protected translate(x: number, y: number) {
    /**
     * Calculates translate coordinates.
     *
     * Indicates dragged y-transformation that's will be updated during the
     * dropping process. Updating Y immediately will effect calculations in
     * transform, that's why it is updated when dragging is done.
     */
    this.translatePlaceholder.setAxes(
      x + this._outerOffset.x,
      y + this._outerOffset.y,
    );

    DFlexBaseElement.transform(
      this.draggedDOM,
      this.translatePlaceholder.x,
      this.translatePlaceholder.y,
    );
  }
}

export default DFlexBaseDraggable;
