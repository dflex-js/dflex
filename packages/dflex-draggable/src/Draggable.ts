import { IDFlexBaseNode } from "@dflex/core-instance";
import type { IPointAxes } from "@dflex/utils";
import store from "./DraggableStoreImp";
import AbstractDraggable from "./AbstractDraggable";

class Draggable extends AbstractDraggable<IDFlexBaseNode> {
  /**
   * Creates an instance of Draggable.
   * Works Only on dragged element level.
   *
   *
   * @param id - elementId
   * @param clickCoordinates -
   */
  constructor(id: string, clickCoordinates: IPointAxes) {
    const element = store.registry[id];

    super(element, clickCoordinates);
  }

  dragAt(x: number, y: number) {
    this.translate(x, y);

    this.draggedElm.translate.clone(this.translatePlaceholder);
  }

  endDragging() {
    this.setDragged(false);
  }
}

export default Draggable;
