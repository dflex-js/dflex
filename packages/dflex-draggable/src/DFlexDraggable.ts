import type { DFlexBaseElement } from "@dflex/core-instance";
import type { AxesPoint } from "@dflex/utils";

import store from "./DFlexDraggableStore";
import DFlexBaseDraggable from "./DFlexBaseDraggable";

class DFlexDraggable extends DFlexBaseDraggable<DFlexBaseElement> {
  /**
   * Creates an instance of Draggable.
   * Works Only on dragged element level.
   *
   * @param id - elementId
   * @param clickCoordinates -
   */
  constructor(id: string, clickCoordinates: AxesPoint) {
    const [element, DOM] = store.getDOMbyElmID(id);
    super(element, DOM, clickCoordinates);

    this._setDOMAttrAndStyle(this.draggedDOM, null, true, false, null, null);
  }

  dragAt(x: number, y: number) {
    this._translate(x, y);

    this.draggedElm._translate._clone(this._translatePlaceholder);
  }

  endDragging() {
    this._setDOMAttrAndStyle(this.draggedDOM, null, false, false, null, null);
  }
}

export default DFlexDraggable;
