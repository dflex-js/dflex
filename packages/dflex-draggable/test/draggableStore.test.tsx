/* eslint-disable import/no-extraneous-dependencies */
import * as React from "react";
import * as ReactTestUtils from "react-dom/test-utils";
import { Root, createRoot } from "react-dom/client";

import { store } from "../src";

describe("Draggable Store", () => {
  const elm0D0 = {
    id: "id-0",
  };

  let container: HTMLDivElement | null;
  let reactRoot: Root;

  beforeAll(() => {
    container = document.createElement("div");
    reactRoot = createRoot(container);
    document.body.appendChild(container);

    function init() {
      const ref = React.createRef<HTMLDivElement>();

      function TestBase() {
        React.useEffect(() => {
          if (ref.current) {
            store.register(elm0D0.id);
          }

          return () => {
            store.unregister(elm0D0.id);
          };
        }, [ref]);

        return <div ref={ref} id={elm0D0.id}></div>;
      }

      ReactTestUtils.act(() => {
        reactRoot.render(<TestBase />);
      });
    }

    init();
  });

  afterAll(() => {
    document.body.removeChild(container!);
    container = null;
    jest.restoreAllMocks();
  });

  it("Registry is not empty", () => {
    expect(store.registry).toBeTruthy();
  });

  it("Registers element and initiates translateX,Y", () => {
    expect(store.registry).toMatchSnapshot();
  });
});
