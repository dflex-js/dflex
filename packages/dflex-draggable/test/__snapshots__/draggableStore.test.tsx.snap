// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draggable Store Registers element and initiates translateX,Y 1`] = `
Map {
  "id-0" => DFlexElement {
    "DOMGrid": PointNum {
      "x": 0,
      "y": 0,
    },
    "DOMOrder": {
      "parent": 0,
      "self": 0,
    },
    "VDOMOrder": {
      "parent": 0,
      "self": 0,
    },
    "_animatedFrame": null,
    "_animation": {
      "duration": "dynamic",
      "easing": "ease-in",
    },
    "_computedDimensions": null,
    "_dragCSS": null,
    "_hasAttribute": Set {},
    "_hasPendingTransform": false,
    "_initialPosition": PointNum {
      "x": 0,
      "y": 0,
    },
    "_isVisible": true,
    "_translateHistory": undefined,
    "depth": 0,
    "id": "id-0",
    "keys": {
      "BK": "dflex_bk_0",
      "CHK": null,
      "PK": "dflex_ky_1_0",
      "SK": "dflex_sk_0_0",
    },
    "readonly": false,
    "rect": BoxRect {
      "bottom": 0,
      "height": 0,
      "left": 0,
      "right": 0,
      "top": 0,
      "width": 0,
    },
    "translate": PointNum {
      "x": 0,
      "y": 0,
    },
  },
}
`;
